libpod/container.go

```diff
@@ -422,14 +446,63 @@ func (c *Container) Start() error {
 	return nil
 }
 
-// Stop stops a container
-func (c *Container) Stop() error {
-	return ErrNotImplemented
+// Stop uses the container's requested stop signal (or SIGTERM if no signal was
+// specified) to stop the container, and if it has not stopped after the given
+// timeout (in seconds), uses SIGKILL to attempt to forcibly stop the container
+// If timeout is 0, SIGKILL will be used immediately
+func (c *Container) Stop(timeout int64) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if err := c.syncContainer(); err != nil {
+		return err
+	}
+
+	if c.state.State == ContainerStateConfigured ||
+		c.state.State == ContainerStateUnknown ||
+		c.state.State == ContainerStatePaused {
+		return errors.Wrapf(ErrCtrStateInvalid, "can only stop created, running, or stopped containers")
+	}
+
+	if err := c.runtime.ociRuntime.stopContainer(c, timeout); err != nil {
+		return err
+	}
+
+	// Sync the container's state to pick up return code
+	if err := c.runtime.ociRuntime.updateContainerStatus(c); err != nil {
```

Comment:
Should you use the sync call?

Reply:
This is getting kind of confusing, because we now have three different calls to sync containers - state.UpdateContainer(), which updates the container from the database; ociRuntime.updateContainerStatus(), which picks up any changes that happened in runc, e.g. the container stopping; and syncContainer(), which combines both. Here we just need to grab the new status from runc, so updateContainerStatus is used. Most of the time, syncContainer() at the start of a function is enough to get the container synced with everything that is going on.